File: 백준/Gold/17471. 게리맨더링/게리맨더링.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.List;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int N;
	static int[] people;
	static List<Integer>[] adjList;
	static boolean[] vis,sel;
	static Queue<Integer> q = new LinkedList<>();
	static int ans = Integer.MAX_VALUE;
	
	public static void main(String[] args) throws IOException {
		N = Integer.parseInt(br.readLine());
		people = new int[N+1];
		vis = new boolean[N+1];
		sel = new boolean[N+1];
		adjList = new ArrayList[N+1];
		for(int i=0; i<=N; i++) adjList[i] = new ArrayList<>();
		st = new StringTokenizer(br.readLine());
		for(int i=1; i<=N; i++) people[i] = Integer.parseInt(st.nextToken());
		for(int i=1; i<=N; i++) {
			st = new StringTokenizer(br.readLine());
			int num = Integer.parseInt(st.nextToken());
			for(int n=0; n<num; n++) {
				int adj=Integer.parseInt(st.nextToken());
				adjList[i].add(adj);
				adjList[adj].add(i);
			}
		}
		divide(1);
		System.out.println(ans==Integer.MAX_VALUE ? -1 : ans);
	}
	
	// 구역 나누기
	static void divide(int idx) {
		if(idx==N+1) {
			if(isConnected()) {
				int group1=0; int group2=0;
				for(int i=1; i<=N; i++) {
					if(sel[i]) group1+=people[i];
					else group2+=people[i];
				}
				ans=Math.min(ans, Math.abs(group1-group2));
			}
			return;
		}
		sel[idx]=true;
		divide(idx+1);
		sel[idx]=false;
		divide(idx+1);
	}
	
	static boolean isConnected() {
		Arrays.fill(vis, false);
		int start1=-1; int start2=-1;
		for(int i=1; i<=N; i++) {
			if(sel[i]) start1=i;
			else start2=i;
		}
		if(start1==-1 || start2==-1) return false;
		bfs(start1,true);
		bfs(start2,false);
		for(int i=1; i<=N; i++) if(!vis[i]) return false;
		return true;
	}
	
	static void bfs(int start, boolean group) {
		q.add(start);
		vis[start]=true;
		while(!q.isEmpty()) {
			int cur = q.poll();
			for(int nxt : adjList[cur]) {
				if(!vis[nxt] && sel[nxt]==group) {
					vis[nxt]=true;
					q.add(nxt);
				}
			}
		}
	}
}

File: 백준/Gold/17471. 게리맨더링/게리맨더링.cc
#include <bits/stdc++.h>
using namespace std;
int N,num,adj,a[11],vis[11],ret=1e9;
vector<vector<int>> v;
void dfs(int cur, bool group[]) {
    vis[cur]=1;
    for(int nxt : v[cur]) {
        if(vis[nxt]) continue;
        if(!group[nxt]) continue;
        dfs(nxt,group);
    }
}
bool check(vector<int> &V) {
    bool group[11]={0};
    memset(vis,0,sizeof(vis));
    for(int i : V) group[i]=1;
    int cnt=0;
    dfs(V[0],group);
    for(int i : V) {
        if(vis[i]) cnt++;
    }
    return cnt==V.size();
}
void calc(vector<int> &A, vector<int> &B) {
    int sumA=0, sumB=0;
    for(int i : A) sumA+=a[i];
    for(int i : B) sumB+=a[i];
    ret=min(ret,abs(sumA-sumB));
}
int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> N;
    for(int i=0; i<=N; i++) v.push_back({});
    for(int i=1; i<=N; i++) cin >> a[i];
    for(int i=1; i<=N; i++) {
        cin >> num;
        for(int j=0; j<num; j++) {
            cin >> adj;
            v[i].push_back(adj);
        }
    }
    for(int i=1; i<(1<<N)-1; i++) {
        vector<int> A,B;
        for(int j=1; j<=N; j++) {
            if(i&(1<<(j-1))) A.push_back(j);
            else B.push_back(j);
        }
        if(check(A)&&check(B)) calc(A,B);
    }
    cout << ((ret==1e9) ? -1 : ret) << '\n';
}